File: javascript/oop-patterns/behavioral-patterns/servant/servant.js
// Слуга предоставляет дополнительную функциональность для обслуживаемых
// объектов
class PrintServant {
  // Распечатать информацию об обслуживаемом объекте
  print(p) {
    console.log(
      `Меня зовут ${p.getName()}, я ${p.getType()} и мне ${p.getAge()} лет.`
    );
  }
}

// Конкретные обслуживаемые классы
class Person {
  #name;
  #age;
  constructor(name, age) {
    this.#name = name;
    this.#age = age;
  }
  getName() {
    return this.#name;
  }
  getAge() {
    return this.#age;
  }
  getType() {
    return 'Человек';
  }
}

class Dog {
  #name;
  #age;
  constructor(name, age) {
    this.#name = name;
    this.#age = age;
  }
  getName() {
    return this.#name;
  }
  getAge() {
    return this.#age;
  }
  getType() {
    return 'Собака';
  }
}

class Cat {
  #name;
  #age;
  constructor(name, age) {
    this.#name = name;
    this.#age = age;
  }
  getName() {
    return this.#name;
  }
  getAge() {
    return this.#age;
  }
  getType() {
    return 'Кот';
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

const p = new Person('Марс', 30);
const d = new Dog('Плутон', 6);
const c = new Cat('Юпитер', 8);

const servant = new PrintServant();

servant.print(p);
servant.print(d);
servant.print(c);

File: go/oop-patterns/behavioral-patterns/servant/servant.go
package main

import "fmt"

// Слуга предоставляет дополнительную функциональность для обслуживаемых
// объектов, при этом он знает только интерфейс обслуживаемых объектов и
// может вызывать соответствующие методы
type printServant struct {
}

// Распечатать информацию об обслуживаемом объекте
func (ps *printServant) print(p printable) {
	fmt.Printf("Меня зовут %v, я %v и мне %v лет.\n", p.getName(), p.getType(), p.getAge())
}

// Интерфейс указывает на то, какие методы должны быть определены обслуживаемыми
// классами
type printable interface {
	getName() string
	getAge() int
	getType() string
}

// Конкретные обслуживаемые типы, реализующие интерфейс
type person struct {
	name string
	age  int
}

func (p *person) getName() string {
	return p.name
}
func (p *person) getAge() int {
	return p.age
}
func (p *person) getType() string {
	return "Человек"
}

func newPerson(name string, age int) *person {
	return &person{name, age}
}

type dog struct {
	name string
	age  int
}

func (d *dog) getName() string {
	return d.name
}
func (d *dog) getAge() int {
	return d.age
}
func (d *dog) getType() string {
	return "Собака"
}

func newDog(name string, age int) *dog {
	return &dog{name, age}
}

type cat struct {
	name string
	age  int
}

func (c *cat) getName() string {
	return c.name
}
func (c *cat) getAge() int {
	return c.age
}
func (c *cat) getType() string {
	return "Кот"
}

func newCat(name string, age int) *cat {
	return &cat{name, age}
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func main() {
	p := newPerson("Марс", 30)
	d := newDog("Плутон", 6)
	c := newCat("Юпитер", 8)

	servant := printServant{}

	servant.print(p)
	servant.print(d)
	servant.print(c)
}
